RecoLocalCalo/HGCalRecProducers/plugins/alpaka/HGCalSoARecHitsProducer.cc

```diff
@@ -0,0 +1,222 @@
+#include "DataFormats/PortableTestObjects/interface/alpaka/TestDeviceCollection.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/stream/EDProducer.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/EDPutToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/ESGetToken.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaTest/interface/AlpakaESTestRecords.h"
+#include "HeterogeneousCore/AlpakaTest/interface/alpaka/AlpakaESTestData.h"
+#include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h"
+#include "Geometry/HGCalGeometry/interface/HGCalGeometry.h"
+#include "DataFormats/HGCRecHit/interface/HGCRecHitCollections.h"
+
+#include "DataFormats/HGCalReco/interface/HGCalSoARecHitsHostCollection.h"
+#include "DataFormats/HGCalReco/interface/alpaka/HGCalSoARecHitsDeviceCollection.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  class HGCalSoARecHitsProducer : public stream::EDProducer<> {
+  public:
+    HGCalSoARecHitsProducer(edm::ParameterSet const& config)
+        : caloGeomToken_(consumesCollector().esConsumes<CaloGeometry, CaloGeometryRecord>()),
+          detector_(config.getParameter<std::string>("detector")),
+          deviceToken_{produces()},
+          initialized_(false) {
+      hits_token_ = consumes<HGCRecHitCollection>(config.getParameter<edm::InputTag>("recHits"));
+      isNose_ = false;
+      if (detector_ == "HFNose") {
+        isNose_ = true;
+      }
+      maxNumberOfThickIndices_ = config.getParameter<unsigned>("maxNumberOfThickIndices");
+      fcPerEle_ = config.getParameter<double>("fcPerEle");
+      fcPerMip_ = config.getParameter<std::vector<double>>("fcPerMip");
+      nonAgedNoises_ = config.getParameter<std::vector<double>>("noises");
+      ecut_ = config.getParameter<double>("ecut");
+      thicknessCorrection_ = config.getParameter<std::vector<double>>("thicknessCorrection");
+      dEdXweights_ = config.getParameter<std::vector<double>>("dEdXweights");
+    }
+
+    ~HGCalSoARecHitsProducer() override = default;
+
+    void produce(device::Event& iEvent, device::EventSetup const& iSetup) override {
+      edm::Handle<HGCRecHitCollection> hits_h;
+
+      edm::ESHandle<CaloGeometry> geom = iSetup.getHandle(caloGeomToken_);
+      rhtools_.setGeometry(*geom);
+      maxlayer_ = rhtools_.lastLayer(isNose_);
+
+      hits_h = iEvent.getHandle(hits_token_);
+      auto const& hits = *(hits_h.product());
+      computeThreshold();
```

Comment:
Can this be called in the constructor, avoiding the need of `initialized_` ?

Reply:
Unfortunately, inside `computeThreshold()` we use some information derived from the geometry fetched via the `EventSetup`.

Reply:
I see.

I assume that information guaranteed to be constant during a job ?

Reply:
OK, if I read the code correctly, the only parameter that depends on the geometry (and thus on the event setup) is `maxlayer_` ?

What happens if the value of `maxlayer_` is inconsistent with `kHGCalLayers` used in `RecoLocalCalo/HGCalRecProducers/plugins/alpaka/HGCalLayerClustersAlgoWrapper.dev.cc` ?

If `maxlayer_` is taken from the geometry, shouldn't `maxNumberOfThickIndices_` also come from the geometry ?

Reply:
If `maxlayer_` is not synchronized with `HGCaLayers`, it could lead to issues. However, as long as `HGCaLayers` is greater than or equal to `maxlayer_`, everything should work fine. I don't anticipate HGCAL changing the number of layers in the geometry.

Regarding `maxNumberOfThickIndices_`, it might be sensible to fetch this from the geometry as well. Currently, this information isn't directly stored in the geometry. This isn't unique to these new producers; the configuration approach for the clustering algorithm is effectively the same as the legacy algorithm.

Reply:
OK, so here is a _suggestion_:
  - make `maxlayer_` a configuration parameter
  - move `computeThreshold()` to the constructor
  - add a check that the value is consistent with the EventSetup
  - add a check that the value is less than or equal to `kHGCalLayers` (` / 2` ?)
 
?